File: frontend/src/types/analytics.ts
export interface OverviewData {
  total_cases: number;
  open_cases: number;
  closed_cases: number;
  filed_cases: number;
  clearance_rate: number;
  avg_resolution_days: number;
  period: {
    from_date: string;
    to_date: string;
  };
  recent_activity?: any[];
}

export interface TrendItem {
  month: string;
  total: number;
  open: number;
  closed: number;
}

export interface DistrictItem {
  district: string;
  count: number;
}

export interface DistributionItem {
  crime_type: string;
  count: number;
  percentage: number;
}

File: backend/models/analytics.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class PeriodInfo(BaseModel):
    from_date: str
    to_date: str


class OverviewResponse(BaseModel):
    total_cases: int
    open_cases: int
    closed_cases: int
    filed_cases: int
    clearance_rate: float = Field(default=0.0, ge=0.0, le=100.0)
    avg_resolution_days: float = Field(default=0.0, ge=0.0)
    period: PeriodInfo


class DistributionItem(BaseModel):
    crime_type: str
    count: int
    percentage: float = Field(default=0.0, ge=0.0, le=100.0)


class TrendItem(BaseModel):
    month: str
    total: int
    open: int
    closed: int


class DistrictItem(BaseModel):
    district: str
    count: int


class HeatMapItem(BaseModel):
    case_id: str
    crime_type: str
    status: str
    date_filed: str
    district: str
    latitude: float
    longitude: float
    intensity: float = 1.0


class StatusBreakdown(BaseModel):
    status: str
    count: int
    percentage: float = Field(default=0.0, ge=0.0, le=100.0)


class AnalyticsResponse(BaseModel):
    overview: OverviewResponse
    crime_distribution: List[DistributionItem] = []
    trends: List[TrendItem] = []
    district_breakdown: List[DistrictItem] = []
    status_breakdown: List[StatusBreakdown] = []
